hazelcast/src/test/java/com/hazelcast/internal/dynamicconfig/ConfigSearchTest.java

```diff
@@ -0,0 +1,1098 @@
+package com.hazelcast.internal.dynamicconfig;
+
+import com.hazelcast.config.AtomicLongConfig;
+import com.hazelcast.config.AtomicReferenceConfig;
+import com.hazelcast.config.CacheSimpleConfig;
+import com.hazelcast.config.CardinalityEstimatorConfig;
+import com.hazelcast.config.Config;
+import com.hazelcast.config.CountDownLatchConfig;
+import com.hazelcast.config.DurableExecutorConfig;
+import com.hazelcast.config.ExecutorConfig;
+import com.hazelcast.config.FlakeIdGeneratorConfig;
+import com.hazelcast.config.ListConfig;
+import com.hazelcast.config.LockConfig;
+import com.hazelcast.config.MapConfig;
+import com.hazelcast.config.MultiMapConfig;
+import com.hazelcast.config.PNCounterConfig;
+import com.hazelcast.config.QueueConfig;
+import com.hazelcast.config.ReliableTopicConfig;
+import com.hazelcast.config.ReplicatedMapConfig;
+import com.hazelcast.config.RingbufferConfig;
+import com.hazelcast.config.ScheduledExecutorConfig;
+import com.hazelcast.config.SemaphoreConfig;
+import com.hazelcast.config.SetConfig;
+import com.hazelcast.config.TopicConfig;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.util.UUID;
+
+import static com.hazelcast.spi.properties.GroupProperty.SEARCH_DYNAMIC_CONFIG_FIRST;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class ConfigSearchTest extends HazelcastTestSupport {
+    private static final String STATIC_NAME = "my.custom.data.*";
+    private static final String DYNAMIC_NAME = "my.custom.data.cache";
+
+    private HazelcastInstance hazelcastInstance;
+
+    private void testTemplate(TestCase testCase) {
+        Config staticHazelcastConfig = getConfig();
+        if (testCase.isDynamicFirst()) {
+            staticHazelcastConfig.setProperty(SEARCH_DYNAMIC_CONFIG_FIRST.getName(), "true");
+        }
+        staticHazelcastConfig.setInstanceName(UUID.randomUUID().toString());
+        testCase.addStaticConfig(staticHazelcastConfig);
+        hazelcastInstance = createHazelcastInstance(staticHazelcastConfig);
+        testCase.addDynamicConfig(hazelcastInstance);
+        testCase.asserts();
+    }
+
+    @After
+    public void tearDown() {
+        Hazelcast.shutdownAll();
```

Comment:
This class is sporadically failing because created instances are not being shut down. The reason is that the instance factories being used to create and shutdown instances are not the same. The invocation `Hazelcast.shutdownAll()` will not shut down instances created by `HazelcastTestSupport#createHazelcastInstance(com.hazelcast.config.Config)` (and `HazelcastTestSupport#factory`).
To correct this, you may simply remove the `tearDown` method in this class since [the parent class already cleans up instances created by using it's methods](https://github.com/hazelcast/hazelcast/blob/db912e3e8e8489f357ab755f7374453a10cb26e8/hazelcast/src/test/java/com/hazelcast/test/HazelcastTestSupport.java#L137-L144)

Reply:
Seems like `DynamicConfigSearchOrderTest` might be failing because of the same reason.